File: 0509FibonacciNumber/FibonacciNumber.go
// -*- coding: utf-8 -*-

package main

import "fmt"

func fib1(n int) int {
	if n < 2 {
		return n
	}
	dp := make([]int, n+1)
	dp[0] = 0
	dp[1] = 1
	for i := 2; i < n+1; i++ {
		dp[i] = dp[i-1] + dp[i-2]
	}
	return dp[len(dp)-1]
}

func fib3(n int) int {
	if n < 2 {
		return n
	}
	ans := 0
	prev2 := 1
	prev := 0
	for i := 2; i <= n; i++ {
		ans = prev2 + prev
		prev = prev2
		prev2 = ans
	}
	return ans
}

func main() {
	fmt.Println(fib1(3))
}

File: 0509FibonacciNumber/FibonacciNumber.py
# -*- coding: utf-8 -*-

class FibonacciNumber:
    def fib1(self, n: "int") -> "int":
        if n == 0:
            return 0
        if n == 1:
            return 1
        dp = [0] * (n + 1)
        dp[1] = 1
        for i in range(2, n + 1):
            dp[i] = dp[i - 1] + dp[i - 2]
        return dp[-1]
        ## dp[i] --> i th fib num
        ## dp[i] = dp[i - 1] + dp[i - 2]


    def fib2(self, n: "int") -> "int":
        if n == 0:
            return 0
        if n == 1:
            return 1
        return self.fib2(n - 1) + self.fib2(n - 2)

    def fib3(self, n: "int") -> "int":
        if not n: return 0
        if n == 1: return 1
        ans = 0
        prev = 0
        prev2 = 1
        for _ in range(2, n+1):
            ans = prev + prev2
            prev = prev2
            prev2 = ans
        return ans

def main():
    test = FibonacciNumber()
    print(test.fib1(2))
    print(test.fib1(5))
    print(test.fib1(30))
    print(test.fib2(2))
    print(test.fib2(5))
    print(test.fib2(30))

if __name__ == '__main__':
    main()
